File: android/java/app/src/main/java/ie/equalit/ouinet_examples/android_java/MainActivity.java
package ie.equalit.ouinet_examples.android_java;

import android.util.Log;
import android.widget.EditText;
import android.widget.TextView;
import android.widget.Toast;
import android.view.View;
import androidx.appcompat.app.AppCompatActivity;
import android.os.Bundle;
import ie.equalit.ouinet.Config;
import ie.equalit.ouinet.Ouinet;
import okhttp3.*;

import javax.net.ssl.*;
import java.io.*;
import java.net.InetSocketAddress;
import java.net.Proxy;
import java.net.URI;
import java.net.URISyntaxException;
import java.security.KeyManagementException;
import java.security.KeyStore;
import java.security.KeyStoreException;
import java.security.NoSuchAlgorithmException;
import java.security.cert.Certificate;
import java.security.cert.CertificateException;
import java.security.cert.CertificateFactory;
import java.security.cert.X509Certificate;
import java.util.concurrent.Executors;

public class MainActivity extends AppCompatActivity {
    private Ouinet ouinet;
    private String ouinet_dir;
    private static final String TAG = "OuinetTester";

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        Config config = new Config.ConfigBuilder(this)
                .setCacheType("bep5-http")
                .setTlsCaCertStorePath("file:///android_asset/cacert.pem")
                .setCacheHttpPubKey(BuildConfig.CACHE_PUB_KEY)
                .setInjectorCredentials(BuildConfig.INJECTOR_CREDENTIALS)
                .setInjectorTlsCert(BuildConfig.INJECTOR_TLS_CERT)
                .setLogLevel(Config.LogLevel.DEBUG)
                .setListenOnTcp("127.0.0.1:8888")
                .setFrontEndEp("127.0.0.1:5555")
                //.setDisableOriginAccess(true)
                //.setDisableProxyAccess(true)
                //.setDisableInjectorAccess(true)
                .build();
        ouinet_dir = config.getOuinetDirectory();

        ouinet = new Ouinet(this, config);
        ouinet.start();

        Executors.newFixedThreadPool(1).execute((Runnable) this::updateServiceState);
    }

    private void updateServiceState() {
      TextView ouinetState = (TextView) findViewById(R.id.ouinetStatus);

      while(true)
      {
          try {
              Thread.sleep(1000);
          } catch (InterruptedException e) {
              e.printStackTrace();
          }

          String state = ouinet.getState().toString();
          runOnUiThread(() -> ouinetState.setText("State: " + state));
      }
    }

    public void getURL(View view) {
        EditText editText = (EditText) findViewById(R.id.url);
        TextView logViewer = (TextView) findViewById(R.id.log_viewer);
        String url = editText.getText().toString();

        Toast toast = Toast.makeText(this, "Loading: " + url, Toast.LENGTH_SHORT);
        toast.show();

        OkHttpClient client = getOuinetHttpClient();
        Request request = new Request.Builder()
                .url(url)
                .header("X-Ouinet-Group", getDhtGroup(url))
                .build();

        client.newCall(request).enqueue(new Callback() {
            public void onFailure(Call call, IOException e) {
                e.printStackTrace();
                logViewer.setText(e.toString());
            }

            public void onResponse(Call call, Response response) throws IOException {
                try (ResponseBody responseBody = response.body()) {
                    Headers responseHeaders = response.headers();
                    for (int i = 0, size = responseHeaders.size(); i < size; i++) {
                        System.out.println(responseHeaders.name(i) + ": " + responseHeaders.value(i));
                    }

                    logViewer.setText(responseHeaders.toString());
                }
            }
        });
    }

    private String getDhtGroup(String url) {
        String domain = null;
        try {
            domain = new URI(url).getSchemeSpecificPart();
            domain = domain.replaceAll("^//", "");
        } catch (URISyntaxException e) {
            e.printStackTrace();
        }
        return domain;
    }

    private OkHttpClient getOuinetHttpClient() {
        try {
            TrustManager[] trustManagers = getOuinetTrustManager();

            OkHttpClient.Builder builder = new OkHttpClient.Builder();
            builder.sslSocketFactory(
                    getSSLSocketFactory(trustManagers),
                    (X509TrustManager) trustManagers[0]);

            // Proxy to ouinet service
            Proxy ouinetService= new Proxy(Proxy.Type.HTTP, new InetSocketAddress("127.0.0.1", 8888));
            builder.proxy(ouinetService);
            return builder.build();
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    private SSLSocketFactory getSSLSocketFactory(TrustManager[] trustManagers) throws NoSuchAlgorithmException, KeyManagementException {
        final SSLContext sslContext = SSLContext.getInstance("TLS");
        sslContext.init(null, trustManagers, new java.security.SecureRandom());
        return sslContext.getSocketFactory();
    }

    private TrustManager[] getOuinetTrustManager() throws NoSuchAlgorithmException, KeyStoreException, CertificateException, IOException {
        return new TrustManager[]{new OuinetTrustManager()};
    }

    private class OuinetTrustManager implements X509TrustManager {
        private X509TrustManager trustManager;
        private Certificate ca;

        public OuinetTrustManager() throws NoSuchAlgorithmException, KeyStoreException, CertificateException, IOException {
            TrustManagerFactory tmf = TrustManagerFactory.getInstance(TrustManagerFactory.getDefaultAlgorithm());
            tmf.init(getKeyStore());

            for (TrustManager tm : tmf.getTrustManagers()) {
                if (tm instanceof X509TrustManager) {
                    trustManager = (X509TrustManager) tm;
                    break;
                }
            }
        }

        private KeyStore getKeyStore() throws KeyStoreException, CertificateException, NoSuchAlgorithmException, IOException {
            KeyStore keyStore = KeyStore.getInstance(KeyStore.getDefaultType());
            keyStore.load(null, null);
            keyStore.setCertificateEntry("ca", getCertificateAuthority());
            return keyStore;
        }

        private Certificate getCertificateAuthority() throws CertificateException {
            InputStream caInput = null;
            try {
                caInput = new FileInputStream(ouinet_dir + "/ssl-ca-cert.pem");
            } catch (FileNotFoundException e) {
                e.printStackTrace();
            }

            CertificateFactory cf = CertificateFactory.getInstance("X.509");
            ca = cf.generateCertificate(caInput);
            return ca;
        }

        @Override
        public void checkClientTrusted(X509Certificate[] chain, String authType) throws CertificateException {
        }

        @Override
        public void checkServerTrusted(X509Certificate[] chain, String authType) throws CertificateException {
            for (X509Certificate cert : chain) {
                Log.d(TAG, "Server Cert Issuer: " + cert.getIssuerDN().getName() + " " + cert.getSubjectDN().getName());
            }
            for (X509Certificate cert : trustManager.getAcceptedIssuers()) {
                Log.d(TAG, "Client Trusted Issuer: " + cert.getIssuerDN().getName());
            }

            trustManager.checkServerTrusted(chain, authType);
        }

        @Override
        public X509Certificate[] getAcceptedIssuers() {
            return new X509Certificate[]{(X509Certificate) ca};
        }

    }
}

File: android/kotlin/app/src/main/java/ie/equalit/ouinet_examples/android_kotlin/MainActivity.kt
package ie.equalit.ouinet_examples.android_kotlin

import android.os.Bundle
import android.util.Log
import android.view.View
import android.widget.Button
import android.widget.EditText
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import ie.equalit.ouinet.Config
import ie.equalit.ouinet.Ouinet
import okhttp3.*
import java.io.FileInputStream
import java.io.FileNotFoundException
import java.io.IOException
import java.io.InputStream
import java.net.InetSocketAddress
import java.net.Proxy
import java.net.URI
import java.net.URISyntaxException
import java.security.*
import java.security.cert.Certificate
import java.security.cert.CertificateException
import java.security.cert.CertificateFactory
import java.security.cert.X509Certificate
import java.util.concurrent.Executors
import javax.net.ssl.*

class MainActivity : AppCompatActivity() {
    private lateinit var ouinet: Ouinet
    lateinit var ouinetDir: String
    private val TAG = "OuinetTester"

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        val get = findViewById<Button>(R.id.get)
        get.setOnClickListener{ getURL(get) }

        var config = Config.ConfigBuilder(this)
            .setCacheType("bep5-http")
            .setTlsCaCertStorePath("file:///android_asset/cacert.pem")
            .setCacheHttpPubKey(BuildConfig.CACHE_PUB_KEY)
            .setInjectorCredentials(BuildConfig.INJECTOR_CREDENTIALS)
            .setInjectorTlsCert(BuildConfig.INJECTOR_TLS_CERT)
            .build()
        ouinetDir = config.ouinetDirectory

        ouinet = Ouinet(this, config)
        ouinet.start()

        Executors.newFixedThreadPool(1).execute(Runnable { this.updateOuinetState() })
    }

    private fun updateOuinetState() {
        val ouinetState = findViewById<View>(R.id.status) as TextView
        while (true) {
            try {
                Thread.sleep(1000)
            } catch (e: InterruptedException) {
                e.printStackTrace()
            }
            val state = ouinet.state.toString()
            runOnUiThread { ouinetState.text = "State: $state" }
        }
    }

    fun getURL(view: View?) {
        val editText = findViewById<View>(R.id.url) as EditText
        val logViewer = findViewById<View>(R.id.log_viewer) as TextView
        val url = editText.text.toString()
        val toast = Toast.makeText(this, "Loading: $url", Toast.LENGTH_SHORT)
        toast.show()

        val client: OkHttpClient = getOuinetHttpClient()
        val request: Request = Request.Builder()
            .url(url)
            .header("X-Ouinet-Group", getDhtGroup(url))
            .build()


        client.newCall(request).enqueue(object : Callback {
            override fun onFailure(call: Call, e: IOException) {
                e.printStackTrace()
                runOnUiThread { logViewer.text = e.toString() }
            }

            @Throws(IOException::class)
            override fun onResponse(call: Call, response: Response) {
                response.body.use { _ ->
                    val responseHeaders = response.headers
                    var i = 0
                    val size = responseHeaders.size
                    while (i < size) {
                        println(responseHeaders.name(i) + ": " + responseHeaders.value(i))
                        i++
                    }
                    runOnUiThread { logViewer.text = responseHeaders.toString() }
                }
            }
        })
    }

    private fun getDhtGroup(url: String): String {
        var domain: String = ""
        try {
            domain = URI(url).schemeSpecificPart
            domain = domain.replace("^//".toRegex(), "")
        } catch (e: URISyntaxException) {
            e.printStackTrace()
        }
        return domain
    }

    private fun getOuinetHttpClient(): OkHttpClient {
        return try {
            val trustManagers: Array<TrustManager> = getOuinetTrustManager()

            val builder = OkHttpClient.Builder()
            builder.sslSocketFactory(
                getSSLSocketFactory(trustManagers),
                (trustManagers[0] as X509TrustManager)
            )

            // Proxy to ouinet service
            val ouinetService = Proxy(Proxy.Type.HTTP, InetSocketAddress("127.0.0.1", 8077))
            builder.proxy(ouinetService)
            return builder.build()
        } catch (e: Exception) {
            throw RuntimeException(e)
        }
    }

    @Throws(NoSuchAlgorithmException::class, KeyManagementException::class)
    private fun getSSLSocketFactory(trustManagers: Array<TrustManager>): SSLSocketFactory {
        val sslContext = SSLContext.getInstance("TLS")
        sslContext.init(null, trustManagers, SecureRandom())
        return sslContext.socketFactory
    }

    @Throws(
        NoSuchAlgorithmException::class,
        KeyStoreException::class,
        CertificateException::class,
        IOException::class
    )
    private fun getOuinetTrustManager(): Array<TrustManager> {
        return arrayOf(OuinetTrustManager())
    }

    inner private class OuinetTrustManager : X509TrustManager {
        private var trustManager: X509TrustManager? = null
        private var ca: Certificate? = null

        init {
            val tmf = TrustManagerFactory.getInstance(TrustManagerFactory.getDefaultAlgorithm())
            tmf.init(keyStore)
            for (tm in tmf.trustManagers) {
                if (tm is X509TrustManager) {
                    trustManager = tm
                    break
                }
            }
        }

        @get:Throws(
            KeyStoreException::class,
            CertificateException::class,
            NoSuchAlgorithmException::class,
            IOException::class
        )
        private val keyStore: KeyStore
            private get() {
                val keyStore = KeyStore.getInstance(KeyStore.getDefaultType())
                keyStore.load(null, null)
                keyStore.setCertificateEntry("ca", certificateAuthority)
                return keyStore
            }

        @get:Throws(CertificateException::class)
        private val certificateAuthority: Certificate?
            private get() {
                var caInput: InputStream? = null
                try {
                    caInput = FileInputStream(ouinetDir + "/ssl-ca-cert.pem")
                } catch (e: FileNotFoundException) {
                    e.printStackTrace()
                }
                val cf = CertificateFactory.getInstance("X.509")
                ca = cf.generateCertificate(caInput)
                return ca
            }

        @Throws(CertificateException::class)
        override fun checkClientTrusted(chain: Array<X509Certificate>, authType: String) {
        }

        @Throws(CertificateException::class)
        override fun checkServerTrusted(chain: Array<X509Certificate>, authType: String) {
            for (cert in chain) {
                Log.d(TAG, "Server Cert Issuer: " + cert.issuerDN.name + " " + cert.subjectDN.name)
            }
            for (cert in trustManager!!.acceptedIssuers) {
                Log.d(TAG, "Client Trusted Issuer: " + cert.issuerDN.name)
            }
            trustManager!!.checkServerTrusted(chain, authType)
        }

        override fun getAcceptedIssuers(): Array<X509Certificate> {
            return arrayOf(ca as X509Certificate)
        }
    }
}